// @rehaps/rehaps-web-components/dist/collection/components/basic/cards/card/card.js
import { h } from "@stencil/core";
export class RwcCard {
  constructor() {
    this.renderShadow = true;
    this.renderBorder = false;
    this.animateBorder = false;
    this.pointer = false;
  }
  render() {
    return (h("div", { class: {
        container: true,
        'animate-border': this.animateBorder,
        border: this.renderBorder,
        shadow: this.renderShadow,
        pointer: this.pointer
      } }, h("slot", null)));
  }
  static get is() { return "rwc-card"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["card.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["card.css"]
    };
  }
  static get properties() {
    return {
      "renderShadow": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set rendering of box-shadow"
        },
        "attribute": "render-shadow",
        "reflect": true,
        "defaultValue": "true"
      },
      "renderBorder": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set rendering of border"
        },
        "attribute": "render-border",
        "reflect": true,
        "defaultValue": "false"
      },
      "animateBorder": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Animate border on hover"
        },
        "attribute": "animate-border",
        "reflect": true,
        "defaultValue": "false"
      },
      "pointer": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Change cursor to pointer on hover"
        },
        "attribute": "pointer",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
}
