// @rehaps/rehaps-web-components/dist/collection/components/navigation/navigation.js
import { h, Host } from "@stencil/core";
export class RwcNavigation {
  constructor() {
    this.type = 'horizontal';
  }
  render() {
    return (h(Host, null, h("div", { class: {
        container: true,
        [this.type]: true
      } }, h("slot", { name: "before" }), h("slot", null), h("slot", { name: "after" }))));
  }
  static get is() { return "rwc-navigation"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["./navigation.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["navigation.css"]
    };
  }
  static get properties() {
    return {
      "type": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "NavigationType",
          "resolved": "\"accordion\" | \"horizontal\" | \"vertical\"",
          "references": {
            "NavigationType": {
              "location": "import",
              "path": "../../interfaces/components/basic/navigation/navigation",
              "id": "src/interfaces/components/basic/navigation/navigation.ts::NavigationType"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The type of the navigation"
        },
        "attribute": "type",
        "reflect": true,
        "defaultValue": "'horizontal'"
      }
    };
  }
}
