// @rehaps/rehaps-web-components-vue/lib/components.ts
/* eslint-disable */
/* tslint:disable */
/* auto-generated vue proxies */
import { defineContainer } from './vue-component-lib/utils';

import type { JSX } from '@rehaps/rehaps-web-components';




export const RwcAccordion = /*@__PURE__*/ defineContainer<JSX.RwcAccordion>('rwc-accordion', undefined, [
  'open',
  'renderDivider',
  'renderIcon'
]);


export const RwcAutoLayout = /*@__PURE__*/ defineContainer<JSX.RwcAutoLayout>('rwc-auto-layout', undefined);


export const RwcButton = /*@__PURE__*/ defineContainer<JSX.RwcButton>('rwc-button', undefined, [
  'iconRight',
  'showIconRight',
  'iconLeft',
  'showIconLeft',
  'type',
  'theme',
  'width',
  'ripple',
  'renderShadow'
]);


export const RwcCard = /*@__PURE__*/ defineContainer<JSX.RwcCard>('rwc-card', undefined, [
  'renderShadow',
  'renderBorder',
  'animateBorder',
  'pointer'
]);


export const RwcContainer = /*@__PURE__*/ defineContainer<JSX.RwcContainer>('rwc-container', undefined, [
  'scrollIntoAnimation',
  'scrollOutAnimation',
  'animateOpacity',
  'animationDuration',
  'visibilityThreshold'
]);


export const RwcDropdown = /*@__PURE__*/ defineContainer<JSX.RwcDropdown>('rwc-dropdown', undefined, [
  'openMode',
  'position',
  'size',
  'mode',
  'dropdownOpen',
  'dropdownClose'
]);


export const RwcFoldingSquares = /*@__PURE__*/ defineContainer<JSX.RwcFoldingSquares>('rwc-folding-squares', undefined);


export const RwcGrid = /*@__PURE__*/ defineContainer<JSX.RwcGrid>('rwc-grid', undefined, [
  'columns',
  'mode',
  'alignment',
  'breakOnHostWidth'
]);


export const RwcHeroBanner = /*@__PURE__*/ defineContainer<JSX.RwcHeroBanner>('rwc-hero-banner', undefined, [
  'src',
  'horizontalCenter',
  'verticalCenter',
  'renderGradient',
  'gradientDirectionBreakpoint',
  'autoHeight',
  'customStructure'
]);


export const RwcIconTextSteps = /*@__PURE__*/ defineContainer<JSX.RwcIconTextSteps>('rwc-icon-text-steps', undefined, [
  'columns',
  'autoUpdateLayout'
]);


export const RwcInfoCard = /*@__PURE__*/ defineContainer<JSX.RwcInfoCard>('rwc-info-card', undefined, [
  'icon',
  'renderShadow',
  'renderBorder',
  'animateBorder',
  'pointer',
  'mode'
]);


export const RwcInput = /*@__PURE__*/ defineContainer<JSX.RwcInput>('rwc-input', undefined, [
  'name',
  'autoComplete',
  'focused',
  'value',
  'placeholder',
  'type',
  'validators',
  'manualValidation',
  'disabled',
  'label'
]);


export const RwcModal = /*@__PURE__*/ defineContainer<JSX.RwcModal>('rwc-modal', undefined, [
  'rwcKey',
  'innerScroll',
  'renderHeaderDivider',
  'renderFooterDivider',
  'renderCloseIcon',
  'animationSpeed',
  'modalOpen',
  'modalClose'
]);


export const RwcNavigation = /*@__PURE__*/ defineContainer<JSX.RwcNavigation>('rwc-navigation', undefined, [
  'type'
]);


export const RwcNavigationElement = /*@__PURE__*/ defineContainer<JSX.RwcNavigationElement>('rwc-navigation-element', undefined, [
  'link',
  'openMode',
  'subNavigationType',
  'subNavigationPosition',
  'renderDividers',
  'renderIcon',
  'preventNavigation',
  'iconLeft',
  'iconRight',
  'navigate'
]);


export const RwcPostCard = /*@__PURE__*/ defineContainer<JSX.RwcPostCard>('rwc-post-card', undefined, [
  'avatar',
  'roundedAvatar',
  'renderShadow',
  'renderBorder',
  'animateBorder',
  'pointer'
]);


export const RwcSelect = /*@__PURE__*/ defineContainer<JSX.RwcSelect>('rwc-select', undefined, [
  'icon',
  'pointer',
  'value',
  'text'
]);


export const RwcSidebar = /*@__PURE__*/ defineContainer<JSX.RwcSidebar>('rwc-sidebar', undefined, [
  'defaultState',
  'rwcKey',
  'position',
  'innerScroll',
  'sidebarOpen',
  'sidebarClose'
]);


export const RwcSlides = /*@__PURE__*/ defineContainer<JSX.RwcSlides>('rwc-slides', undefined, [
  'rwcKey',
  'defaultSlide',
  'autoSlide',
  'autoSlideDelay',
  'indexChange'
]);


export const RwcSpinner = /*@__PURE__*/ defineContainer<JSX.RwcSpinner>('rwc-spinner', undefined);


export const RwcSystemNotifications = /*@__PURE__*/ defineContainer<JSX.RwcSystemNotifications>('rwc-system-notifications', undefined);


export const RwcTabs = /*@__PURE__*/ defineContainer<JSX.RwcTabs>('rwc-tabs', undefined, [
  'rwcKey',
  'mode',
  'defaultTab',
  'indexChange'
]);


export const RwcTag = /*@__PURE__*/ defineContainer<JSX.RwcTag>('rwc-tag', undefined, [
  'icon',
  'renderIcon',
  'selected',
  'clickable',
  'deleteTag'
]);


export const RwcUserCard = /*@__PURE__*/ defineContainer<JSX.RwcUserCard>('rwc-user-card', undefined, [
  'avatar',
  'roundedAvatar',
  'renderShadow',
  'renderBorder',
  'animateBorder',
  'locationIcon',
  'ratingIcon',
  'pointer'
]);
