// @rehaps/rehaps-web-components/dist/collection/components/basic/cards/post-card/post-card.js
import { h, Host } from "@stencil/core";
import AccountIcon from "@material-design-icons/svg/outlined/account_circle.svg";
export class RwcPostCard {
  constructor() {
    this.avatar = '';
    this.roundedAvatar = false;
    this.renderShadow = true;
    this.renderBorder = false;
    this.animateBorder = false;
    this.pointer = false;
  }
  render() {
    return (h(Host, null, h("rwc-card", { class: "card", renderShadow: this.renderShadow, renderBorder: this.renderBorder, animateBorder: this.animateBorder, pointer: this.pointer }, h("div", { class: "container" }, h("div", { class: "user" }, h("div", { class: "avatar" }, h("slot", { name: "avatar" }, this.renderAvatar())), h("div", { class: "name" }, h("slot", { name: "name" }))), h("div", { class: "text" }, h("div", { class: "heading content-spacing" }, h("slot", { name: "heading" })), h("div", { class: "description content-spacing" }, h("slot", { name: "description" })), h("div", { class: "footer content-spacing" }, h("slot", { name: "footer" })))))));
  }
  renderAvatar() {
    if (this.avatar) {
      return (h("img", { src: this.avatar, alt: "avatar", class: {
          icon: true,
          rounded: this.roundedAvatar
        } }));
    }
    return (h("div", { innerHTML: AccountIcon, class: "svg-icon" }));
  }
  static get is() { return "rwc-post-card"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["post-card.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["post-card.css"]
    };
  }
  static get properties() {
    return {
      "avatar": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Path to the avatar."
        },
        "attribute": "avatar",
        "reflect": true,
        "defaultValue": "''"
      },
      "roundedAvatar": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set avatar rounded."
        },
        "attribute": "rounded-avatar",
        "reflect": false,
        "defaultValue": "false"
      },
      "renderShadow": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set rendering of box-shadow."
        },
        "attribute": "render-shadow",
        "reflect": true,
        "defaultValue": "true"
      },
      "renderBorder": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set rendering of border."
        },
        "attribute": "render-border",
        "reflect": true,
        "defaultValue": "false"
      },
      "animateBorder": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Animate border on hover."
        },
        "attribute": "animate-border",
        "reflect": true,
        "defaultValue": "false"
      },
      "pointer": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Change cursor to pointer on hover."
        },
        "attribute": "pointer",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
}
