// @rehaps/rehaps-web-components/dist/components/rwc-hero-banner.js
import { proxyCustomElement, HTMLElement, getAssetPath, h, Host, Fragment } from '@stencil/core/internal/client';

const heroBannerCss = "@font-face{font-family:\"Inter\";font-style:normal;font-weight:400;font-display:swap;src:url(\"assets/fonts/inter/Inter-Regular.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Regular.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:500;font-display:swap;src:url(\"assets/fonts/inter/Inter-Medium.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Medium.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:600;font-display:swap;src:url(\"assets/fonts/inter/Inter-SemiBold.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-SemiBold.woff?v=1.0\") format(\"woff\")}:host{--hero-banner-text-color:#fff;--hero-banner-font-family:Inter, system-ui, sans-serif;--hero-banner-container-padding:2rem;--hero-banner-background-repeat:no-repeat;--hero-banner-background-size:cover;--hero-banner-min-height:31.25rem;--hero-banner-gradient-direction:90deg;--hero-banner-gradient-background-image:linear-gradient(var(--hero-banner-gradient-direction), #111111, rgba(17, 17, 17, 0));--hero-banner-content-grid-column-count:4;--hero-banner-content-grid-column:span 3;--hero-banner-content-grid-column--mobile:span 4;--hero-banner-content-justify-content:center;display:block}:host(.vertical){--hero-banner-gradient-direction:0deg;--hero-banner-content-justify-content:flex-end}@media all and (min-width: 30rem){.auto-height{--hero-banner-min-height:35rem}}@media all and (min-width: 64rem){.auto-height{--hero-banner-min-height:26rem}}.container{background-repeat:var(--hero-banner-background-repeat);background-size:var(--hero-banner-background-size);color:var(--hero-banner-text-color);display:flex;font-family:var(--hero-banner-font-family);flex-direction:column;min-height:var(--hero-banner-min-height);position:relative}.container:not(.custom){padding:var(--hero-banner-container-padding)}.gradient{position:absolute;top:0;left:0;height:100%;width:100%;background-image:var(--hero-banner-gradient-background-image);pointer-events:none}.grid{display:grid;grid-template-columns:repeat(var(--hero-banner-content-grid-column-count), 1fr);flex-grow:1;z-index:1}.grid .content{grid-column:var(--hero-banner-content-grid-column--mobile);display:flex;flex-direction:column;justify-content:var(--hero-banner-content-justify-content);z-index:1}@media all and (min-width: 48rem){.grid .content{grid-column:var(--hero-banner-content-grid-column)}}";

const RwcHeroBanner$1 = proxyCustomElement(class RwcHeroBanner extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.src = getAssetPath('assets/images/banners/Banner_example.jpg');
    this.horizontalCenter = 50;
    this.verticalCenter = 50;
    this.renderGradient = false;
    this.gradientDirectionBreakpoint = 0;
    this.autoHeight = true;
    this.customStructure = false;
    this.vertical = false;
  }
  get hostWidth() {
    let hostWidth = this.host.offsetWidth;
    if (typeof hostWidth !== "number") {
      hostWidth = window.innerWidth;
    }
    return hostWidth;
  }
  componentDidLoad() {
    this.updateGradientMode();
  }
  onWindowResize() {
    this.updateGradientMode();
  }
  updateGradientMode() {
    if (this.gradientDirectionBreakpoint > 0) {
      this.vertical = (this.hostWidth >= this.gradientDirectionBreakpoint ? false : true);
    }
  }
  render() {
    return (h(Host, { class: {
        vertical: this.vertical
      } }, h("div", { class: {
        container: true,
        'custom': this.customStructure,
        'auto-height': this.autoHeight
      }, style: {
        'background-image': `url(${this.src})`,
        'background-position-x': `${this.horizontalCenter}%`,
        'background-position-y': `${this.verticalCenter}%`
      } }, this.renderGradientElement(), this.renderContent(this.customStructure))));
  }
  renderGradientElement() {
    if (this.renderGradient) {
      return (h("div", { class: "gradient" }));
    }
    return h(Fragment, null);
  }
  renderContent(customStructure) {
    if (customStructure) {
      return (h("div", { class: "content" }, h("slot", null)));
    }
    return (h("div", { class: "grid" }, h("div", { class: "content" }, h("slot", null))));
  }
  static get assetsDirs() { return ["../../assets/images/banners"]; }
  get host() { return this; }
  static get style() { return heroBannerCss; }
}, [1, "rwc-hero-banner", {
    "src": [513],
    "horizontalCenter": [2, "horizontal-center"],
    "verticalCenter": [2, "vertical-center"],
    "renderGradient": [4, "render-gradient"],
    "gradientDirectionBreakpoint": [2, "gradient-direction-breakpoint"],
    "autoHeight": [516, "auto-height"],
    "customStructure": [4, "custom-structure"],
    "vertical": [32]
  }, [[9, "resize", "onWindowResize"]]]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["rwc-hero-banner"];
  components.forEach(tagName => { switch (tagName) {
    case "rwc-hero-banner":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, RwcHeroBanner$1);
      }
      break;
  } });
}

const RwcHeroBanner = RwcHeroBanner$1;
const defineCustomElement = defineCustomElement$1;

export { RwcHeroBanner, defineCustomElement };
