// @rehaps/rehaps-web-components/dist/components/rwc-post-card.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { A as AccountIcon } from './account_circle.js';
import { d as defineCustomElement$2 } from './card.js';

const postCardCss = "@font-face{font-family:\"Inter\";font-style:normal;font-weight:400;font-display:swap;src:url(\"assets/fonts/inter/Inter-Regular.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Regular.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:500;font-display:swap;src:url(\"assets/fonts/inter/Inter-Medium.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Medium.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:600;font-display:swap;src:url(\"assets/fonts/inter/Inter-SemiBold.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-SemiBold.woff?v=1.0\") format(\"woff\")}:host{--post-card-border-width:2px;--post-card-border-style:solid;--post-card-border-color:rgb(224, 224, 224);--post-card-hover-border-color:rgb(144, 238, 144);--post-card-border-radius:0.75rem;--post-card-border-transition:border-bottom-color 0.4s,\n    border-left-color 0.4s,\n    border-right-color 0.4s,\n    border-top-color 0.4s,\n    width 0.4s ease-in-out,\n    height 0.4s ease-in-out;--post-card-box-shadow:0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -4px rgba(0, 0, 0, 0.1);--post-card-box-shadow-animated-hover:0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 8px 10px -6px rgba(0, 0, 0, 0.1);--post-card-container-transition:box-shadow 0.4s ease-in-out;--post-card-content-justify-start:start;--post-card-content-justify-end:end;--post-card-content-spacing:0.5rem;--post-card-inner-padding:1rem;--post-card-font-size-lg:1.25rem;--post-card-font-size-md:1rem;--post-card-font-size-sm:0.875rem;--post-card-icon-size:2.5rem;--post-card-icon-margin-right:1rem;--post-card-icon-color:rgb(117, 117, 117);--post-card-gap:0.5rem;font-family:Inter, system-ui, sans-serif;display:block}.card{--card-border-width:var(--post-card-border-width);--card-border-style:var(--post-card-border-style);--card-border-color:var(--post-card-border-color);--card-hover-border-color:var(--post-card-hover-border-color);--card-border-radius:var(--post-card-border-radius);--card-border-transition:var(--post-card-border-transition);--card-box-shadow-animated-hover:var(--post-card-box-shadow-animated-hover);--card-container-transition:var(--post-card-container-transition);--card-inner-padding:var(--post-card-inner-padding);--card-box-shadow:var(--post-card-box-shadow)}.pointer{cursor:pointer}.user{display:flex;align-items:center;justify-content:var(--post-card-content-justify-start)}.name{font-size:var(--post-card-font-size-lg)}.heading{font-weight:bold;font-size:var(--post-card-font-size-md)}.description{font-size:var(--post-card-font-size-sm)}.footer{font-size:var(--post-card-font-size-sm);display:flex;align-items:center;justify-content:var(--post-card-content-justify-end);gap:var(--post-card-gap)}.content-spacing{margin-top:var(--post-card-content-spacing)}.avatar{align-self:center;flex-shrink:0;height:var(--post-card-icon-size);width:var(--post-card-icon-size);margin-right:var(--post-card-icon-margin-right)}.avatar svg{height:100%;width:100%}.avatar path{fill:var(--post-card-icon-color)}.avatar .icon{width:100%;height:100%}.avatar .rounded{border-radius:50%}";

const RwcPostCard$1 = proxyCustomElement(class RwcPostCard extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.avatar = '';
    this.roundedAvatar = false;
    this.renderShadow = true;
    this.renderBorder = false;
    this.animateBorder = false;
    this.pointer = false;
  }
  render() {
    return (h(Host, null, h("rwc-card", { class: "card", renderShadow: this.renderShadow, renderBorder: this.renderBorder, animateBorder: this.animateBorder, pointer: this.pointer }, h("div", { class: "container" }, h("div", { class: "user" }, h("div", { class: "avatar" }, h("slot", { name: "avatar" }, this.renderAvatar())), h("div", { class: "name" }, h("slot", { name: "name" }))), h("div", { class: "text" }, h("div", { class: "heading content-spacing" }, h("slot", { name: "heading" })), h("div", { class: "description content-spacing" }, h("slot", { name: "description" })), h("div", { class: "footer content-spacing" }, h("slot", { name: "footer" })))))));
  }
  renderAvatar() {
    if (this.avatar) {
      return (h("img", { src: this.avatar, alt: "avatar", class: {
          icon: true,
          rounded: this.roundedAvatar
        } }));
    }
    return (h("div", { innerHTML: AccountIcon, class: "svg-icon" }));
  }
  static get style() { return postCardCss; }
}, [1, "rwc-post-card", {
    "avatar": [513],
    "roundedAvatar": [4, "rounded-avatar"],
    "renderShadow": [516, "render-shadow"],
    "renderBorder": [516, "render-border"],
    "animateBorder": [516, "animate-border"],
    "pointer": [4]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["rwc-post-card", "rwc-card"];
  components.forEach(tagName => { switch (tagName) {
    case "rwc-post-card":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, RwcPostCard$1);
      }
      break;
    case "rwc-card":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const RwcPostCard = RwcPostCard$1;
const defineCustomElement = defineCustomElement$1;

export { RwcPostCard, defineCustomElement };
