// @rehaps/rehaps-web-components/dist/components/rwc-navigation-element.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host, Fragment } from '@stencil/core/internal/client';
import { E as ExpandMore, d as defineCustomElement$4 } from './accordion.js';
import { d as defineCustomElement$3 } from './dropdown.js';
import { d as defineCustomElement$2 } from './navigation.js';

const ChevronRight = `<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M9.29 6.71a.996.996 0 0 0 0 1.41L13.17 12l-3.88 3.88a.996.996 0 1 0 1.41 1.41l4.59-4.59a.996.996 0 0 0 0-1.41L10.7 6.7c-.38-.38-1.02-.38-1.41.01z"/></svg>`;

const navigationElementCss = "@font-face{font-family:\"Inter\";font-style:normal;font-weight:400;font-display:swap;src:url(\"assets/fonts/inter/Inter-Regular.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Regular.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:500;font-display:swap;src:url(\"assets/fonts/inter/Inter-Medium.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Medium.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:600;font-display:swap;src:url(\"assets/fonts/inter/Inter-SemiBold.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-SemiBold.woff?v=1.0\") format(\"woff\")}:host{display:block;--navigation-element-text-color:rgb(33, 33, 33);--navigation-element-text-hover-color:rgb(33, 33, 33);--navigation-element-text-active-color:rgb(33, 33, 33);--navigation-element-text-decoration:none;--navigation-element-hover-text-decoration:none;--navigation-element-active-text-decoration:none;--navigation-element-background-color:unset;--navigation-element-hover-background-color:unset;--navigation-element-active-background-color:unset;--navigation-element-font-family:Inter, system-ui, sans-serif;--navigation-element-font-size:1rem;--navigation-element-font-weight:500;--navigation-element-line-height:1.2;--navigation-element-link-padding:0.25rem 0.5rem;--navigation-element-link-display:flex;--navigation-element-divider-width:1px;--navigation-element-divider-style:solid;--navigation-element-divider-border-color:rgb(224, 224, 224);--navigation-element-sub-navigation-dropdown-container-padding:0.5rem 0;--navigation-element-sub-navigation-dropdown-border:1px solid rgb(224, 224, 224);--navigation-element-sub-navigation-dropdown-border-radius:0.5rem;--navigation-element-sub-navigation-dropdown-item-padding:0.125rem 0.5rem;--navigation-element-sub-navigation-dropdown-item-background-color:transparent;--navigation-element-sub-navigation-dropdown-item-hover-background-color:rgb(238, 238, 238);--navigation-element-sub-navigation-dropdown-min-width:6.25rem;--navigation-element-sub-navigation-container-padding:0;--navigation-element-sub-navigation-accordion-heading-font-family:Inter, system-ui, sans-serif;--navigation-element-sub-navigation-accordion-heading-font-size:1.125rem;--navigation-element-sub-navigation-accordion-heading-font-weight:400;--navigation-element-sub-navigation-accordion-heading-line-height:1.3;--navigation-element-sub-navigation-accordion-heading-padding:0.5rem 0;--navigation-element-sub-navigation-accordion-content-padding:0 0 0 0.25rem;--navigation-element-sub-navigation-accordion-divider-background-color:rgb(224, 224, 224);--navigation-element-sub-navigation-accordion-divider-height:1px;--navigation-element-right-icon-open-transform:rotate(-180deg);--navigation-element-right-icon-transition:transform var(--navigation-element-animation-duration) ease-in-out;--navigation-element-animation-duration:200ms;display:flex;flex-direction:column;justify-content:center}.container{font-family:var(--navigation-element-font-family);font-size:var(--navigation-element-font-size);font-weight:var(--navigation-element-font-weight);line-height:var(--navigation-element-line-height);padding:var(--navigation-element-container-padding)}.container:not(.divider-none){border-color:var(--navigation-element-divider-border-color)}.container.divider-bottom{border-bottom-width:var(--navigation-element-divider-width);border-bottom-style:var(--navigation-element-divider-style)}.container.divider-right{border-right-width:var(--navigation-element-divider-width);border-right-style:var(--navigation-element-divider-style)}.link{background-color:var(--navigation-element-background-color);color:var(--navigation-element-text-color);display:var(--navigation-element-link-display);flex-direction:row;justify-content:space-between;padding:var(--navigation-element-link-padding);transition:background-color var(--navigation-element-link-style-transition-duration) linear, color var(--navigation-element-link-style-transition-duration) linear}.link:hover{background-color:var(--navigation-element-hover-background-color);color:var(--navigation-element-text-hover-color)}.link:hover .text{text-decoration:var(--navigation-element-hover-text-decoration)}.link:active{background-color:var(--navigation-element-active-background-color);color:var(--navigation-element-text-active-color)}.link:active .text{text-decoration:var(--navigation-element-active-text-decoration)}.icon{display:flex;gap:var(--navigation-element-link-gap);flex-direction:column;justify-content:center;fill:rgb(189, 189, 189)}.icon-left{margin-right:var(--navigation-element-link-gap)}.icon-right{margin-left:var(--navigation-element-link-gap);transition:var(--navigation-element-right-icon-transition)}.text{display:flex;flex-direction:column;justify-content:center;text-decoration:var(--navigation-element-text-decoration)}a{cursor:pointer;text-decoration:inherit;color:inherit}.sub-navigation-dropdown{min-width:var(--navigation-element-sub-navigation-dropdown-min-width);--dropdown-container-padding:var(--navigation-element-sub-navigation-dropdown-container-padding);--dropdown-border:var(--navigation-element-sub-navigation-dropdown-border);--dropdown-border-radius:var(--navigation-element-sub-navigation-dropdown-border-radius);--dropdown-animation-duration:var(--navigation-element-animation-duration);--dropdown-item-padding:var(--navigation-element-sub-navigation-dropdown-item-padding);--dropdown-item-background-color:transparent;--dropdown-item-hover-background-color:rgb(238, 238, 238)}.sub-navigation-accordion{--accordion-default-font-family:var(--navigation-element-sub-navigation-accordion-heading-font-family);--accordion-heading-font-size:var(--navigation-element-sub-navigation-accordion-heading-font-size);--accordion-heading-font-weight:var(--navigation-element-sub-navigation-accordion-heading-font-weight);--accordion-content-line-height:var(--navigation-element-sub-navigation-accordion-heading-line-height);--accordion-heading-padding:var(--navigation-element-sub-navigation-accordion-heading-padding);--accordion-content-padding:var(--navigation-element-sub-navigation-accordion-content-padding);--accordion-animation-duration:var(--navigation-element-animation-duration);--accordion-divider-background-color:var(--navigation-element-sub-navigation-accordion-divider-background-color);--accordion-divider-height:var(--navigation-element-sub-navigation-accordion-divider-height)}.sub-navigation-accordion.has-sub-navigation{--accordion-heading-font-weight:var(--navigation-element-font-weight)}.sub-navigation{--navigation-container-padding:var(--navigation-element-sub-navigation-container-padding)}.dropdown-open .icon.rotate{transform:var(--navigation-element-right-icon-open-transform)}";

const RwcNavigationElement$1 = proxyCustomElement(class RwcNavigationElement extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.navigate = createEvent(this, "navigate", 7);
    this.link = RwcNavigationElement.NO_NAVIGATION_HREF;
    this.openMode = 'hover';
    this.subNavigationType = 'vertical';
    this.subNavigationPosition = 'bottom';
    this.renderDividers = false;
    this.renderIcon = true;
    this.preventNavigation = false;
    this.iconLeft = '';
    this.iconRight = '';
    this.dropdownOpen = false;
  }
  get hasSubNavigation() {
    return Boolean(this.host.querySelector('[slot="sub-navigation"]'));
  }
  onDropdownOpen(event) {
    if (event.target === this.host) {
      this.dropdownOpen = true;
    }
  }
  onDropdownClose(event) {
    if (event.target === this.host) {
      this.dropdownOpen = false;
    }
  }
  componentWillLoad() {
    if (this.subNavigationType === 'accordion') {
      this.openMode = 'manual';
    }
  }
  handleNavigation(event) {
    if (this.preventNavigation) {
      event.preventDefault();
    }
    const closestAnchor = event.composedPath()[0].closest('a');
    if (closestAnchor) {
      this.navigate.emit(closestAnchor.href);
    }
  }
  getNavigationElementHref() {
    if (this.openMode === 'hover' &&
      this.subNavigationType !== 'accordion' ||
      !this.hasSubNavigation) {
      return this.link;
    }
    return RwcNavigationElement.NO_NAVIGATION_HREF;
  }
  render() {
    return (h(Host, null, h("div", { class: {
        container: true,
        [this.openMode]: true,
        divider: this.renderDividers,
        [`sub-nav-type-${this.subNavigationType}`]: true,
        'dropdown-open': this.dropdownOpen
      } }, this.renderNavigationElement())));
  }
  renderNavigationElement() {
    if (this.subNavigationType === 'accordion') {
      return (h("rwc-accordion", { class: {
          'sub-navigation-accordion': true,
          'has-sub-navigation': this.hasSubNavigation
        }, renderDivider: this.renderDividers, open: false, renderIcon: this.renderIcon && this.hasSubNavigation }, this.renderLinkElement({ slot: 'heading' })));
    }
    return this.renderLinkElement({ slot: null });
  }
  renderLinkElement(params) {
    return (h(Fragment, null, h("a", { href: this.getNavigationElementHref(),
      'slot': params.slot || '', onClick: (event) => this.handleNavigation(event) }, h("div", { class: "link" }, h("slot", { name: "before" }, this.renderLeftIcon()), h("div", { class: "text" }, h("slot", null)), h("slot", { name: "after" }, this.renderRightIcon()))), this.renderSubNavigationContainer()));
  }
  renderSubNavigationContainer() {
    if (this.hasSubNavigation) {
      if (this.subNavigationType === 'accordion') {
        return (h("rwc-navigation", { class: "sub-navigation", type: this.subNavigationType }, h("slot", { name: "sub-navigation" })));
      }
      else {
        return (h("rwc-dropdown", { class: "sub-navigation-dropdown", openMode: this.openMode, mode: 'fixed', position: this.subNavigationPosition, size: "auto" }, h("slot", { name: "sub-navigation" })));
      }
    }
    return h(Fragment, null);
  }
  renderLeftIcon() {
    if (this.iconLeft) {
      return (h("img", { src: this.iconLeft, alt: "Left icon", class: "icon icon icon-left" }));
    }
    return h(Fragment, null);
  }
  renderRightIcon() {
    if (this.iconRight) {
      return (h("img", { src: this.iconRight, alt: "Right icon", class: {
          icon: true,
          ['icon-right']: true,
          rotate: this.subNavigationPosition === 'bottom' && this.subNavigationType !== 'accordion'
        } }));
    }
    else if (this.subNavigationType === 'vertical' && this.renderIcon && this.hasSubNavigation) {
      return (h("div", { innerHTML: this.subNavigationPosition === 'side' ? ChevronRight : ExpandMore, class: {
          icon: true,
          ['icon-right']: true,
          rotate: this.subNavigationPosition === 'bottom'
        } }));
    }
  }
  get host() { return this; }
  static get style() { return navigationElementCss; }
}, [1, "rwc-navigation-element", {
    "link": [513],
    "openMode": [1537, "open-mode"],
    "subNavigationType": [513, "sub-navigation-type"],
    "subNavigationPosition": [513, "sub-navigation-position"],
    "renderDividers": [4, "render-dividers"],
    "renderIcon": [4, "render-icon"],
    "preventNavigation": [4, "prevent-navigation"],
    "iconLeft": [1, "icon-left"],
    "iconRight": [1, "icon-right"],
    "dropdownOpen": [32]
  }, [[0, "dropdownOpen", "onDropdownOpen"], [0, "dropdownClose", "onDropdownClose"]]]);
RwcNavigationElement$1.NO_NAVIGATION_HREF = 'javascript:void(0)';
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["rwc-navigation-element", "rwc-accordion", "rwc-dropdown", "rwc-navigation"];
  components.forEach(tagName => { switch (tagName) {
    case "rwc-navigation-element":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, RwcNavigationElement$1);
      }
      break;
    case "rwc-accordion":
      if (!customElements.get(tagName)) {
        defineCustomElement$4();
      }
      break;
    case "rwc-dropdown":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "rwc-navigation":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const RwcNavigationElement = RwcNavigationElement$1;
const defineCustomElement = defineCustomElement$1;

export { RwcNavigationElement, defineCustomElement };
