// @rehaps/rehaps-web-components/dist/esm/rwc-slides.entry.js
import { r as registerInstance, c as createEvent, h, a as Host, g as getElement } from './index-e9508d80.js';
import { E as EventManager } from './eventManager-f421930c.js';
import { R as RWCActionHelper } from './rwcActionHelper-202b4497.js';

const slidesCss = "@font-face{font-family:\"Inter\";font-style:normal;font-weight:400;font-display:swap;src:url(\"assets/fonts/inter/Inter-Regular.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Regular.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:500;font-display:swap;src:url(\"assets/fonts/inter/Inter-Medium.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Medium.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:600;font-display:swap;src:url(\"assets/fonts/inter/Inter-SemiBold.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-SemiBold.woff?v=1.0\") format(\"woff\")}:host{--slides-animation-duration:200ms;--slides-slide-fade-transition:opacity var(--slides-animation-duration) ease-in-out,\n    visibility 0ms linear var(--slides-animation-duration),\n    position 0ms linear var(--slides-animation-duration),\n    left 0ms linear var(--slides-animation-duration);--slides-slide-active-fade-transition:opacity var(--slides-animation-duration) ease-in-out var(--slides-animation-duration),\n    visibility 0ms linear var(--slides-animation-duration),\n    position 0ms linear var(--slides-animation-duration),\n    left 0ms linear var(--slides-animation-duration);display:block}.container{width:100%;overflow-x:hidden;position:relative}.slide-flex{display:flex;flex-direction:row}.slide-wrapper{width:100%}::slotted(*){left:-99999px;opacity:0;position:absolute;transition:var(--slides-slide-fade-transition);visibility:hidden}::slotted(.active){left:0;opacity:1;position:unset;transition:var(--slides-slide-active-fade-transition);visibility:visible}";

const RwcSlides = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.changeEvent = createEvent(this, "indexChange", 7);
    this.eventManager = new EventManager();
    this.actions = ['goTo'];
    this.autoSlideCoroutine = null;
    this.rwcKey = RWCActionHelper.getDefaultComponentKey('slides');
    this.defaultSlide = 0;
    this.autoSlide = false;
    this.autoSlideDelay = 5000;
    this.activeSlide = 0;
  }
  onAutoSlideChange() {
    this.init();
  }
  onMouseEnter() {
    clearInterval(this.autoSlideCoroutine);
  }
  onMouseLeave() {
    if (this.autoSlide) {
      this.initAutoSlide();
    }
  }
  validateTargetIndex(value) {
    if (value > this.slides.length - 1 || value < 0) {
      return 0;
    }
    return value;
  }
  componentDidLoad() {
    this.init();
  }
  disconnectedCallback() {
    clearInterval(this.autoSlideCoroutine);
    this.eventManager.clearEventListeners();
  }
  init() {
    this.eventManager.clearEventListeners();
    this.slides = this.slideContainer.assignedElements();
    this.initDefaultSlide();
    this.initGoToEventListeners();
    clearInterval(this.autoSlideCoroutine);
    if (this.autoSlide) {
      this.initAutoSlide();
    }
  }
  initDefaultSlide() {
    var _a;
    this.activeSlide = this.validateTargetIndex(this.activeSlide);
    if (((_a = this.slides) === null || _a === void 0 ? void 0 : _a.length) > 0) {
      this.slides[this.activeSlide].classList.add(RwcSlides.ACTIVE_SLIDE_CLASS);
    }
  }
  initAutoSlide() {
    clearInterval(this.autoSlideCoroutine);
    this.autoSlideCoroutine = setInterval(() => {
      this.goTo(this.validateTargetIndex(this.activeSlide + 1));
    }, this.autoSlideDelay);
  }
  async goTo(targetIndex) {
    const parsedIndex = this.validateTargetIndex(targetIndex instanceof CustomEvent ? Number.parseInt(targetIndex.detail.next) : targetIndex);
    if (this.activeSlide !== parsedIndex) {
      this.slides.forEach(slide => {
        slide.classList.remove(RwcSlides.ACTIVE_SLIDE_CLASS);
      });
      const originalEvent = targetIndex instanceof CustomEvent ? targetIndex : null;
      this.setActiveSlide(parsedIndex, originalEvent);
    }
  }
  setActiveSlide(targetIndex, originalEvent) {
    var _a;
    this.slides[targetIndex].classList.add(RwcSlides.ACTIVE_SLIDE_CLASS);
    this.changeEvent.emit((_a = originalEvent === null || originalEvent === void 0 ? void 0 : originalEvent.detail) !== null && _a !== void 0 ? _a : { origin: this.host, next: targetIndex, previous: this.activeSlide });
    this.activeSlide = targetIndex;
  }
  initGoToEventListeners() {
    RWCActionHelper.bindRWCActions(this, this.host, this.actions, this.rwcKey, 'indexChange');
  }
  render() {
    return (h(Host, null, h("div", { class: {
        container: true
      } }, h("div", { class: "slide-wrapper" }, h("div", { class: "slide-flex" }, h("slot", { ref: (slot) => this.slideContainer = slot, onSlotchange: () => this.init() }))))));
  }
  get host() { return getElement(this); }
  static get watchers() { return {
    "autoSlide": ["onAutoSlideChange"]
  }; }
};
RwcSlides.ACTIVE_SLIDE_CLASS = 'active';
RwcSlides.style = slidesCss;

export { RwcSlides as rwc_slides };
