// @rehaps/rehaps-web-components/dist/esm/rwc-accordion_2.entry.js
import { r as registerInstance, h, a as Host, g as getElement, F as Fragment } from './index-e9508d80.js';
import { E as ExpandMore } from './expand_more-fa2ad8bf.js';

const accordionCss = "@font-face{font-family:\"Inter\";font-style:normal;font-weight:400;font-display:swap;src:url(\"assets/fonts/inter/Inter-Regular.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Regular.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:500;font-display:swap;src:url(\"assets/fonts/inter/Inter-Medium.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Medium.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:600;font-display:swap;src:url(\"assets/fonts/inter/Inter-SemiBold.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-SemiBold.woff?v=1.0\") format(\"woff\")}:host{--accordion-default-font-family:Inter, system-ui, sans-serif;--accordion-heading-padding:0.5rem 0;--accordion-heading-font-family:var(--accordion-default-font-family);--accordion-heading-font-size:1.25rem;--accordion-heading-font-weight:500;--accordion-content-line-height:1.2;--accordion-content-padding:0.5rem 0 1rem 0;--accordion-content-font-family:var(--accordion-default-font-family);--accordion-content-font-size:1rem;--accordion-content-font-weight:400;--accordion-content-line-height:1.2;--accordion-divider-background-color:rgb(224, 224, 224);--accordion-divider-height:1px;--accordion-icon-color:rgb(189, 189, 189);--accordion-icon-open-transform:rotate(-180deg);--accordion-animation-duration:200ms;display:block}.heading{cursor:pointer;user-select:none;font-family:var(--accordion-heading-font-family);font-size:var(--accordion-heading-font-size);font-weight:var(--accordion-heading-font-weight);line-height:var(--accordion-content-line-height);padding:var(--accordion-heading-padding)}.heading.with-icon{display:flex;flex-direction:row;justify-content:space-between}.left,.right{display:flex;flex-direction:column;justify-content:center}.left{flex-grow:1}.right{flex-shrink:0}.icon{display:flex;flex-direction:column;justify-content:center;fill:var(--accordion-icon-color);transition:transform var(--accordion-animation-duration) ease-in-out}.content-wrapper{height:0;overflow:hidden;transition:height var(--accordion-animation-duration) ease-in-out}.content{font-family:var(--accordion-default-font-family);font-size:var(--accordion-content-font-size);font-weight:var(--accordion-content-font-weight);padding:var(--accordion-content-padding)}.divider{background-color:var(--accordion-divider-background-color);height:var(--accordion-divider-height)}.open .content-wrapper{height:auto}.open .icon.rotate{transform:var(--accordion-icon-open-transform)}";

const RwcAccordion = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.open = false;
    this.renderDivider = true;
    this.renderIcon = true;
  }
  get animationDuration() {
    return Number.parseInt(getComputedStyle(this.host).getPropertyValue('--accordion-animation-duration'));
  }
  onHeadingClick() {
    this.setState(!this.open);
  }
  async setState(value) {
    if (this.animationCallback !== null) {
      clearTimeout(this.animationCallback);
    }
    this.open = value;
    if (value) {
      this.contentWrapperElement.style.height = `${this.contentElement.offsetHeight}px`;
    }
    else {
      this.contentWrapperElement.style.height = `${this.contentElement.offsetHeight}px`;
      await new Promise(requestAnimationFrame);
      this.contentWrapperElement.style.height = `0px`;
    }
    this.animationCallback = setTimeout(() => {
      this.contentWrapperElement.style.height = '';
      this.animationCallback = null;
    }, this.animationDuration);
  }
  render() {
    return (h(Host, { class: {
        open: this.open,
      } }, h("div", { class: {
        container: true,
        open: this.open
      } }, h("div", { class: {
        heading: true,
        'with-icon': this.renderIcon
      }, onClick: () => this.onHeadingClick() }, h("div", { class: "left" }, h("slot", { name: "heading" })), this.renderHeadingIconContainer()), h("div", { ref: (element) => this.contentWrapperElement = element, class: "content-wrapper" }, h("div", { ref: (element) => this.contentElement = element, class: "content" }, h("slot", null))), this.renderDividerElement())));
  }
  renderDividerElement() {
    if (this.renderDivider) {
      return h("div", { class: "divider" });
    }
    return h(Fragment, null);
  }
  renderHeadingIconContainer() {
    if (this.renderIcon) {
      return (h("div", { class: "right" }, h("slot", { name: "heading-icon" }, this.renderDefaultIcon())));
    }
    return h(Fragment, null);
  }
  renderDefaultIcon() {
    return h("div", { class: "icon rotate", innerHTML: ExpandMore });
  }
  get host() { return getElement(this); }
};
RwcAccordion.style = accordionCss;

const navigationCss = "@font-face{font-family:\"Inter\";font-style:normal;font-weight:400;font-display:swap;src:url(\"assets/fonts/inter/Inter-Regular.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Regular.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:500;font-display:swap;src:url(\"assets/fonts/inter/Inter-Medium.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-Medium.woff?v=1.0\") format(\"woff\")}@font-face{font-family:\"Inter\";font-style:normal;font-weight:600;font-display:swap;src:url(\"assets/fonts/inter/Inter-SemiBold.woff2?v=1.0\") format(\"woff2\"), url(\"assets/fonts/inter/Inter-SemiBold.woff?v=1.0\") format(\"woff\")}:host{--navigation-container-padding:0.5rem;--navigation-container-border:none;--navigation-container-border-radius:none;--navigation-font-family:Inter, system-ui, sans-serif;--navigation-font-size:1rem;--navigation-font-weight:500;--navigation-line-height:1.2;display:block}.container{display:flex;flex-direction:row;font-family:var(--navigation-font-family);font-size:var(--navigation-font-size);font-weight:var(--navigation-font-weight);line-height:var(--navigation-line-height);padding:var(--navigation-container-padding);border:var(--navigation-container-border);border-radius:var(--navigation-container-border-radius)}.container.vertical,.container.accordion{flex-direction:column}";

const RwcNavigation = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.type = 'horizontal';
  }
  render() {
    return (h(Host, null, h("div", { class: {
        container: true,
        [this.type]: true
      } }, h("slot", { name: "before" }), h("slot", null), h("slot", { name: "after" }))));
  }
};
RwcNavigation.style = navigationCss;

export { RwcAccordion as rwc_accordion, RwcNavigation as rwc_navigation };
