// @rehaps/rehaps-web-components-vue/dist/components.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RwcUserCard = exports.RwcTag = exports.RwcTabs = exports.RwcSystemNotifications = exports.RwcSpinner = exports.RwcSlides = exports.RwcSidebar = exports.RwcSelect = exports.RwcPostCard = exports.RwcNavigationElement = exports.RwcNavigation = exports.RwcModal = exports.RwcInput = exports.RwcInfoCard = exports.RwcIconTextSteps = exports.RwcHeroBanner = exports.RwcGrid = exports.RwcFoldingSquares = exports.RwcDropdown = exports.RwcContainer = exports.RwcCard = exports.RwcButton = exports.RwcAutoLayout = exports.RwcAccordion = void 0;
const utils_1 = require("./vue-component-lib/utils");
exports.RwcAccordion = (0, utils_1.defineContainer)('rwc-accordion', undefined, [
    'open',
    'renderDivider',
    'renderIcon'
]);
exports.RwcAutoLayout = (0, utils_1.defineContainer)('rwc-auto-layout', undefined);
exports.RwcButton = (0, utils_1.defineContainer)('rwc-button', undefined, [
    'iconRight',
    'showIconRight',
    'iconLeft',
    'showIconLeft',
    'type',
    'theme',
    'width',
    'ripple',
    'renderShadow'
]);
exports.RwcCard = (0, utils_1.defineContainer)('rwc-card', undefined, [
    'renderShadow',
    'renderBorder',
    'animateBorder',
    'pointer'
]);
exports.RwcContainer = (0, utils_1.defineContainer)('rwc-container', undefined, [
    'scrollIntoAnimation',
    'scrollOutAnimation',
    'animateOpacity',
    'animationDuration',
    'visibilityThreshold'
]);
exports.RwcDropdown = (0, utils_1.defineContainer)('rwc-dropdown', undefined, [
    'openMode',
    'position',
    'size',
    'mode',
    'dropdownOpen',
    'dropdownClose'
]);
exports.RwcFoldingSquares = (0, utils_1.defineContainer)('rwc-folding-squares', undefined);
exports.RwcGrid = (0, utils_1.defineContainer)('rwc-grid', undefined, [
    'columns',
    'mode',
    'alignment',
    'breakOnHostWidth'
]);
exports.RwcHeroBanner = (0, utils_1.defineContainer)('rwc-hero-banner', undefined, [
    'src',
    'horizontalCenter',
    'verticalCenter',
    'renderGradient',
    'gradientDirectionBreakpoint',
    'autoHeight',
    'customStructure'
]);
exports.RwcIconTextSteps = (0, utils_1.defineContainer)('rwc-icon-text-steps', undefined, [
    'columns',
    'autoUpdateLayout'
]);
exports.RwcInfoCard = (0, utils_1.defineContainer)('rwc-info-card', undefined, [
    'icon',
    'renderShadow',
    'renderBorder',
    'animateBorder',
    'pointer',
    'mode'
]);
exports.RwcInput = (0, utils_1.defineContainer)('rwc-input', undefined, [
    'name',
    'autoComplete',
    'focused',
    'value',
    'placeholder',
    'type',
    'validators',
    'manualValidation',
    'disabled',
    'label'
]);
exports.RwcModal = (0, utils_1.defineContainer)('rwc-modal', undefined, [
    'rwcKey',
    'innerScroll',
    'renderHeaderDivider',
    'renderFooterDivider',
    'renderCloseIcon',
    'animationSpeed',
    'modalOpen',
    'modalClose'
]);
exports.RwcNavigation = (0, utils_1.defineContainer)('rwc-navigation', undefined, [
    'type'
]);
exports.RwcNavigationElement = (0, utils_1.defineContainer)('rwc-navigation-element', undefined, [
    'link',
    'openMode',
    'subNavigationType',
    'subNavigationPosition',
    'renderDividers',
    'renderIcon',
    'preventNavigation',
    'iconLeft',
    'iconRight',
    'navigate'
]);
exports.RwcPostCard = (0, utils_1.defineContainer)('rwc-post-card', undefined, [
    'avatar',
    'roundedAvatar',
    'renderShadow',
    'renderBorder',
    'animateBorder',
    'pointer'
]);
exports.RwcSelect = (0, utils_1.defineContainer)('rwc-select', undefined, [
    'icon',
    'pointer',
    'value',
    'text'
]);
exports.RwcSidebar = (0, utils_1.defineContainer)('rwc-sidebar', undefined, [
    'defaultState',
    'rwcKey',
    'position',
    'innerScroll',
    'sidebarOpen',
    'sidebarClose'
]);
exports.RwcSlides = (0, utils_1.defineContainer)('rwc-slides', undefined, [
    'rwcKey',
    'defaultSlide',
    'autoSlide',
    'autoSlideDelay',
    'indexChange'
]);
exports.RwcSpinner = (0, utils_1.defineContainer)('rwc-spinner', undefined);
exports.RwcSystemNotifications = (0, utils_1.defineContainer)('rwc-system-notifications', undefined);
exports.RwcTabs = (0, utils_1.defineContainer)('rwc-tabs', undefined, [
    'rwcKey',
    'mode',
    'defaultTab',
    'indexChange'
]);
exports.RwcTag = (0, utils_1.defineContainer)('rwc-tag', undefined, [
    'icon',
    'renderIcon',
    'selected',
    'clickable',
    'deleteTag'
]);
exports.RwcUserCard = (0, utils_1.defineContainer)('rwc-user-card', undefined, [
    'avatar',
    'roundedAvatar',
    'renderShadow',
    'renderBorder',
    'animateBorder',
    'locationIcon',
    'ratingIcon',
    'pointer'
]);
